common/persistence/nosql/nosql_history_store_test.go

```diff
@@ -527,47 +527,630 @@ func TestDeleteHistoryBranch_unusedBranch(t *testing.T) {
+// |	                                   ┼
+// |	                         ┌─────────▼─────────┐
+// |	                         │     Tree: 123     │
+// |	                         │     Branch: A     │
+// |	                         │     Node: 2       │
+// |	                         └─────────┬─────────┘
+// |	          ┌────────────────────────┘
+// |	          │
+// |	┌─────────▼─────────┐
+// |	│     Tree: 123     │
+// |	│     Branch: C     │
+// |	│     Node: 3       │
+// |	└─────────┬─────────┘
+// |	          │
+// |	          │
+// |	┌─────────▼─────────┐
+// |	│     Tree: 123     │
+// |	│     Branch: C     │
+// |	│     Node: 4       │
+// |	└───────────────────┘
+func TestDeleteHistoryBranch_DeletedAncestor(t *testing.T) {
+	store, dbMock, _ := setUpMocks(t)
+
+	request := &persistence.InternalDeleteHistoryBranchRequest{
+		BranchInfo: types.HistoryBranch{
+			TreeID:   "TestTreeID",
+			BranchID: "B",
+			Ancestors: []*types.HistoryBranchRange{
+				{
+					BranchID:    "A",
+					BeginNodeID: 0,
+					EndNodeID:   2,
+				},
+			},
 		},
+		ShardID: testShardID,
+	}
+
+	expectedTreeFilter := &nosqlplugin.HistoryTreeFilter{
+		ShardID:  testShardID,
+		TreeID:   "TestTreeID",
+		BranchID: common.Ptr("B"),
+	}
+
+	expectedNodeFilters := []*nosqlplugin.HistoryNodeFilter{
 		{
 			ShardID:   testShardID,
 			TreeID:    "TestTreeID",
-			BranchID:  "TestAncestorBranchID",
-			MinNodeID: 7,
+			BranchID:  "B",
+			MinNodeID: 2,
+		},
+		{
+			ShardID:   testShardID,
+			TreeID:    "TestTreeID",
+			BranchID:  "A",
+			MinNodeID: 3,
 		},
 	}
 
-	// Expect to delete the history branch
-	dbMock.EXPECT().DeleteFromHistoryTreeAndNode(gomock.Any(), expecedTreeFilter, expectedNodeFilters).
+	dbMock.EXPECT().DeleteFromHistoryTreeAndNode(gomock.Any(), expectedTreeFilter, expectedNodeFilters).
 		Return(nil).Times(1)
 
 	historyTree := []*nosqlplugin.HistoryTreeRow{
 		{
 			ShardID:  testShardID,
 			TreeID:   "TestTreeID",
-			BranchID: "TestBranchID",
+			BranchID: "B",
 			Ancestors: []*types.HistoryBranchRange{
 				{
 					BranchID:    "TestAncestorBranchID",
 					BeginNodeID: 0,
-					EndNodeID:   7,
+					EndNodeID:   2,
+				},
+			},
+		},
+		{
+			ShardID:  testShardID,
+			TreeID:   "TestTreeID",
+			BranchID: "C",
+			Ancestors: []*types.HistoryBranchRange{
+				{
+					BranchID:    "A",
+					BeginNodeID: 0,
+					EndNodeID:   3,
+				},
+			},
+		},
+		// notably A does not exist
+	}
+
+	dbMock.EXPECT().SelectFromHistoryTree(gomock.Any(), gomock.Any()).
+		Return(historyTree, nil).Times(1)
+
+	err := store.DeleteHistoryBranch(ctx.Background(), request)
+	assert.NoError(t, err)
+}
+
+// |  In this scenario, something like the following has happened:
+// |  There was a normal, original workflow, which was then branched (perhaps by a reset).
+// |  By the time the child workflow is being cleaned up, the parent workflow has been deleted already,
+// |  and does not exist in the history_tree table as a valid branch.
```

Comment:
(also irl, but)
probably worth avoiding "child workflow" entirely, and sticking with "branch[ed] workflow", given the naming collision